### DISCUSSION TITLE
Complete Transition to Rust: NautilusTrader

### DISCUSSION THREAD

**Discussion by xtay315** [nautechsystems/nautilus_trader#1582] [Q&A] [answered]
Hello everyone,

I just arrived at the Nautilus Trader community and have looked into the project. I noticed that important underlying data definitions, storage, and networking have been ported to Rust. However, there still exists a significant amount of FFI and PyO3 interoperability in the project. Is there a timeline for when the project plans to fully migrate to Rust and Python?

Thank you.

**Comment by cjdsellers** [comment, marked as answer]
Hi @xtay315

There's a few of us working on this incrementally with varying amounts of bandwidth. The goal certainly is to eliminate Cython and all FFI from the NautilusTrader codebase - currently there's just too much variance to give any estimates though. Rust can take a surprisingly long time to satisfy the compiler with ported code, and often design patterns which worked in an OO and Python context don't work well with Rust.

Let me know if the Rust codebase is something you're interested in becoming involved with though.

**Comment by xtay315** [comment]
Thank you for your response, I really admire your contributions, this is a great code repository, good job.

**Comment by huracosunah** [comment]
Does the Nautilus team intend on making this work completely in Rust? Meaning no Python layer. Write and run a strategy in Rust. I'd be interesting in helping with that.

**Reply by cjdsellers** [reply]
Hi @huracosunah 

Thanks for reaching out. The goal is for backtesting and live trading to be possible in pure Rust, although there will also be a Python API available via optional bindings and a Python package (there's a `python` feature flag which can just be turned off at compile time for a Rust only system).

Currently it's possible to run actors for data processing live in Rust, soon execution will also be possible, then we'll move to complete the backtesting pieces.

I hope that helps to clarify, and your contributions are welcome!
